message view: Use inline date separators.

showell opened the pull request:
There may be a simpler way to check czo here, but the
main goal here is to just deploy-test the one-liner to
get some immediate user feedback.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
OK, test-deploying now.

timabbott commented:
See https://chat.zulip.org/#narrow/stream/101-design/subject/date.20separators/near/661609.  I also found a weird bug where `EDITED` appeared somewhere totally weird:

![image](https://user-images.githubusercontent.com/2746074/48519532-198cf380-e822-11e8-89be-5cec9ae3ec9a.png)

timabbott commented:
So I think we have support for the concept but a handful of bugs reported by multiple people:
* We should still be showing sender headers for messages on a new date.
* The thing I posted above with EDITED positioning.
* Floating recipient bar should update when you scroll past a date separator.

timabbott commented:
(Merged the second refactoring commit, now that I have a bit more time)

timabbott commented:
(See also https://github.com/zulip/zulip/issues/10839)

showell commented:
I rebased, but I haven't addressed any of the feedback yet.

showell commented:
The first three commits here are refactoring commits, and they all probably make sense regardless of how the experiment turns out.

I wasn't able to reproduce the `EDITED` things, but I think I may have fixed it as part of re-working my changes, plus there was some plausibly related recent activity on master that I picked up in rebases.  Tim, you should definitely test this out, since you reported the bug and may have some particular behavior that triggers it.

The **last two commits** change the actual UI here, and they can arguably be squashed together.  The first commit puts the date separators **inside** groups, not between them.  Then the second commit fixes the problem that @jackrzhang reported in #10839, and we now update the FRB as we scroll.

I made a couple decisions here that I think are pretty defensible, but we'll see if it elicits feedback:

* I now **always** show dates in recipient bars.  I think this makes sense in general, but it particularly makes sense in a world where use the slimmer date separators within groups that span multiple days.
* I **don't** put date separators between groups.  I think the recipient bars, with the dates, make them a bit redundant.  For the first group, you can visually scan up to either a date separator inside, or the date will show in the recipient bar (and if it's floating, it will be there, too).  And then for the second group, there will be a date in the recipient bar.

showell commented:
![image](https://user-images.githubusercontent.com/142908/50484635-595d0400-09a6-11e9-9823-da08edccf7d8.png)

timabbott commented:
I merged the 3 prep commits and push back to this PR.

timabbott commented:
> I now always show dates in recipient bars. I think this makes sense in general, but it particularly makes sense in a world where use the slimmer date separators within groups that span multiple days.

I think this decision we'll likely want to undo, in that we did the suppression of dates there when they are irrelevant as part of an effort to declutter the UI, and that change was popular at the time.  

Changing this just requires bringing back the `show_date` value and setting it correctly, right?

timabbott commented:
@rishig can you play around with this and let me know if you can find any bugs and what you think about the UX?  It seems worth doing a round of that before test-deploying on chat.zulip.org.

rishig commented:
cool! 
* I occasionally get browser errors of the form `source_recip_bar is not defined` while trying to navigate with up/down arrows over the recipient bars (couldn't figure out a repro, but has happened twice while testing).

* Getting this when reaching top of narrow (pressing up arrow doesn't fix it, but reload does):
![image](https://user-images.githubusercontent.com/890911/50671254-cef45000-0ff6-11e9-98fd-c20d779d9e71.png)

There are probably a number of styling changes we'll want to do, but I think the most noticable one is that the bar feels uncentered; i.e. too much bottom-margin relative to top-margin.

Only if it's not too hard, we could also do the following before testing on czo:
* Have only 1 date (the new one)
* Move it to the right side, in line with the other dates and timestamps
* The date can be a bit bolder, and use sentence-case, similar to the date in the recipient bar

showell commented:
@rishig Thanks for testing this!  It's unfortunate that you were testing off a broken version.  I think some rebasing glitch or last-minute cosmetic cleanup introduced the "source_recip_bar" error, which not only caused actual errors but prevented the floating recipient bar code from working.  (It was basically source_recip_bar -> source_recipient_bar, and a spot was missed).

I re-pushed with a fix.

I agree on these changes:

- Showing just the "forward" date seems fine to me (and right-aligning it, bolding it).
- Fix the vertical centering.

I'm not really sure what you mean by "sentence casing" for the dates--don't we already do that?

Here's another screenshot:

![image](https://user-images.githubusercontent.com/142908/50697087-17f7d480-0ff6-11e9-8b02-f2c722353137.png)

showell commented:
(Just to be clear, I didn't try to address any of the styling concerns in the latest push--I just rebased and fixed the var-naming glitch I mentioned in the previous comment.)

showell commented:
The date dividers are mostly driven by this code in `static/js/timerender.js`:

~~~ js
137 // Given an XDate object 'time', return a DOM node that initially
138 // displays the human-formatted date, and is updated automatically as
139 // necessary (e.g. changing "Today" to "Yesterday" to "Jul 1").
140 // If two dates are given, it renders them as:
141 // --- ▲ Yesterday ▲ ------ ▼ Today ▼ ---
142
143 // (What's actually spliced into the message template is the contents
144 // of this DOM node as HTML, so effectively a copy of the node. That's
145 // okay since to update the time later we look up the node by its id.)
146 exports.render_date = function (time, time_above, today) {

showell commented:
And then styling is in `static/styles/zulip.scss` (although maybe should be moved, but doesn't have to happen as part of this effort):

~~~ css

2646 .date_row .date-line {
2647     display: inline-block;
2648     vertical-align: middle;
2649     width: 33%;
2650     border-top: 1px solid hsl(0, 0%, 88%);
2651     border-bottom: 1px solid hsl(0, 0%, 100%);
2652     margin: 0px 5px 0px 5px;
2653 }
2654
2655 .sub-unsub-message span,
2656 .date_row span {
2657     display: block;
2658     background: inherit;
2659     padding: 4px;
2660     overflow: hidden;
2661     text-transform: uppercase;
2662     font-size: 0.8em;
2663     opacity: 0.5;
2664 }

rishig commented:
Played around with this; looks good to me! I think ready for pushing to master or czo, and then we can fix-forward on the styling.

timabbott commented:
OK, just test-deployed on chat.zulip.org.

timabbott commented:
Styling-wise, I think we want to remove the `margin-bottom: 10px` on `.date-row` on line 2637 of zulip.scss.

timabbott commented:
I squashed that change in, and added a fixes tag for #10171, and pushed back here.  Test-deploying to czo again.

timabbott commented:
> I'm not really sure what you mean by "sentence casing" for the dates--don't we already do that?

I think this part is about the small caps case used in the date-line thingys.

timabbott commented:
After playing with this a bit, I like the overall experience, so we should definitely merge something in this direction.  I still really want https://github.com/zulip/zulip/pull/10820#issuecomment-451332564 to be changed back. Interleaved views like this (which is the only thing that decision affects) now have the date part look really spammy:

![image](https://user-images.githubusercontent.com/2746074/50718653-b6ac2180-1046-11e9-8396-353a43d49f42.png)

timabbott commented:
OK.  @showell can your refactor this to preserve the existing behavior for the recipient bar date fields (I understand this may require some logic changes)?  I think after that, we'll be ready to merge.

showell commented:
@timabbott and @rishig This is ready for another round of testing.  I de-clutter the dates.

Tim, I have commits split out for helping the review, but they should technically be squashed back together.

De-cluttering the dates motivated a significant rewrite here.  Part of the complications with de-cluttering dates under the new paradigm here is that the FRB for any section can change on the fly.  You could theoretically just update the "next" recipient bar every time, as well as the old FRB's neighbor, but this gets edge-case-y really fast.  Instead, I just recompute the info for all recipient bars that are visible during every `update` call.  (Although I didn't implement it yet, I think we could also speed up compose-fade using the `relevant_recipient_bars` function.)

Technically the `floating_recipient_bar` module has grown in scope here to be more like `recipient_bar`, but the diff's already complicated enough.

timabbott commented:
I think you didn't -- looks OK.  This has been test-deployed on chat.zulip.org for today without report of issues, so I'll try to integrate it first thing tomorrow.

timabbott commented:
@rishig just to update you on this PR,

Tim Abbott: Well, I think the actual design we're talking about here is this:

1. In the current master design, each date value (Today, Yesterday, etc.) appears at most once in the right-hand column, in either a floating recipient bar or a normal recipient bar. This is achieved by breaking recipient bars for each date change, in the static rendering.
2. In the current czo / this branch design, each date value (Today, Yesterday, etc.) appears at most once in the right-hand column, in either a floating recipient bar or a normal recipient bar. This cannot be achieved statically, and so we dynamically make it happen via rerendering when processing floating recipient bar updates. This produces a glitch when scrolling up in the "all messages" or a stream view, where a bunch of `Today` strings flash briefly in the date area for the 50ms until the floating recipient bar logic runs, and then they disappear.  We could potentially make this glitch much less visible by adjusting the default state of these elements in the rendering code to be likely the value they're going to end up at.
3. The proposed alternative design that I think we eventually want to get to requires restyling the date-change bars so that the place they show today's date is in the far-right column where we put timestamps. Once that's done, it would again be possible to statically render the message list's decisions of which recipient bars show a date, with each date appearing at most once, but we'd need somewhat more complicated logic for both the floating recipient bar and this static rendering to take into account date rows in addition to recipient rows.

```
Option 3 would involve changing this:
-------------------------- < June 14 ------ > June 15 ------------------------------

to instead something like this:

-------------------------------------------------------------------------------------------  June 14
```

I think barring us doing that styling change, this model is about the best we can do (though as I mentioned, I think we could choose the hide/show state for dates in the initial rendering better, such that rather than every message starting with the date appearing and that disappearing 50ms after you scroll the message into view, they instead start out with something more similar to the old static decisions for whether to display dates, and auto-correct that -- that would make the glitch like 30x less visible)

rishig commented:
Makes sense to me. I think the dates not appearing at all for 50ms is also probably fine. 
(And yeah, looking forward to the styling change of option 3).

timabbott commented:
OK, I finished cleaning this up and merged it.  @showell FYI, I fixed a few things:
* Left in place the between-recipient rows date dividers, since for the moment those are net helpful.
* Fixed `candidate_recipient_bars` to handle there being non-recipient-row things between them.  This required debugging the annoying fact that `.prev(selector)` returns nothing if the selector doesn't apply (not the next child that matches!).
* Adjusted the date row selection logic in `first_visible` to handle two tricky corner cases where we were either displaying no date at all in the floating recipient bar incorrectly or displaying a confusing date (when the floating recipient bar partially obscures a message feed recipient bar).

Huge thanks for building this @showell!  I think it makes zulip's UI look a lot cleaner.
